### PR TITLE
FEATURE: Export chat messages to CSV file

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
To export chat messages, go to `/admin/plugins/chat` and click the <kbd>Create export</kbd> button in the _Export chat messages_ section:

Screenshot 2023-06-20 at 23 22 06

You'll receive a direct message when the export is finished:

Screenshot 2023-06-14 at 22 23 02

Currently, this exports all messages from the last 6 months, but not more than 10000 messages.

This exports _all_ chat messages, including messages from private channels and users' direct conversations. This also exports messages that were deleted.

Exported data looks like this:

| id | chat_channel_id | chat_channel_name | user_id | username | message | cooked | created_at | updated_at | deleted_at | in_reply_to_id | last_editor_id | last_editor_username |
|---|---|---|---|---|---|---|---|---|---|---|---|---|
| 1 | 1 | Staff | 1 | admin1 | Hey | `<p>Hey</p>` | 2023-02-03 16:34:19 UTC | 2023-04-26 16:58:18 UTC | | | 1 | admin1 |

### CODE REVIEW

**Comment by AndrewPrigorshnev** [comment]
Tests fails because our custom `eq_time` matcher doesn't handle nil values. I'll take care of that.

**Comment by AndrewPrigorshnev** [comment]
> Tests fails because our custom `eq_time` matcher doesn't handle nil values. I'll take care of that.

Here it is: https://github.com/discourse/discourse/pull/22116.

File: plugins/chat/assets/javascripts/discourse/templates/admin-plugins-chat.hbs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+<AdminExportMessages />
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/templates/admin-plugins-chat.hbs:1]: Probably need to be prefixed with Chat at least here `<Chat::Admin::ExportMessages />`

File: plugins/chat/config/routes.rb [hunk 2]
```diff
@@ -41,6 +41,10 @@
     get "/channels/:channel_id/summarize" => "summaries#get_summary"
   end
 
+  namespace :admin, defaults: { format: :json, constraints: StaffConstraint.new } do
+    post "export/messages" => "export#export_messages"
```
Comment (jjaffeux) [inline plugins/chat/config/routes.rb:45]: I would argue that `export_messages#create` would be better, but up to you

File: plugins/chat/lib/chat/messages_exporter.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          username
+          message
+          cooked
+          created_at
+          updated_at
+          deleted_at
+          in_reply_to_id
+          last_editor_id
+          last_editor_username
+        ]
```
Comment (jjaffeux) [inline plugins/chat/lib/chat/messages_exporter.rb:46]: I wonder if we couldn't find a way to refactor this in a way where we a mapping defined at the top level, and on one function we use the keys and the other, the values. This way we ensure we are in sync and we don't forget to change one place when we change the other.

Comment (jjaffeux) [reply plugins/chat/lib/chat/messages_exporter.rb:46]: Works for me :+1:

File: plugins/chat/config/routes.rb (hunk 2 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/config/routes.rb:45]: I'd prefer to leave it like this, this way in the future we'll be able to add other exports to this route / controller.

File: plugins/chat/assets/javascripts/discourse/templates/admin-plugins-chat.hbs (hunk 1 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/assets/javascripts/discourse/templates/admin-plugins-chat.hbs]: Fixed.

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**